core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java

```
@@ -93,53 +113,64 @@
-                                final Optional<Backoff> backoffOnExceptionOpt = retryStrategy().shouldRetry(
-                                        rootReqDuplicator.duplicateStream(), cause);
-                                if (backoffOnExceptionOpt.isPresent()) {
-                                    retry0(ctx, backoffOnExceptionOpt.get(), rootReqDuplicator, action,
-                                           deferredRes, -1);
-                                } else { // exception that is not for retry occurred
-                                    deferredRes.close(cause);
-                                    rootReqDuplicator.close();
-                                }
-                            } else { // normal response
-                                deferredRes.delegate(resDuplicator.duplicateStream(true));
-                                rootReqDuplicator.close();
-                            }
-                        }));
-                    }
-                }));
+        final ContentPreviewResponse contentPreviewResponse =
+                new ContentPreviewResponse(resDuplicator.duplicateStream(), contentPreviewLength);
+        retryStrategy().shouldRetry(rootReqDuplicator.duplicateStream(), contentPreviewResponse)
+                       .handle(voidFunction((backoffOpt, unused) -> {
+                           if (backoffOpt.isPresent()) {
+                               final long millisAfter =
+                                       useRetryAfter ? getRetryAfterMillis(resDuplicator.duplicateStream())
+                                                     : -1;
+                               resDuplicator.close();
+
+                               long nextDelay;
+                               try {
+                                   nextDelay = getNextDelay(ctx, backoffOpt.get(), millisAfter);
+                                   if (nextDelay < 0) { // exceed the number of max attempt
```

How about throwing `RetryGiveUpException` in `getNextDelay` instead of returning `-1` so that we can remove this `if` block?

---

That's a great idea~!! I will change it.